File: src/agent/orchestrator/decompose.ts
/**
 * Subtask Decomposition
 * 
 * Breaks tasks into implementable subtasks to prevent "one-shot" failures.
 * Uses heuristics and optionally LLM for complex task decomposition.
 */
import { Effect } from "effect";
import * as fs from "node:fs";
import * as path from "node:path";
import type { Task } from "../../tasks/index.js";
import {
  type Subtask,
  type SubtaskList,
  getSubtasksPath,
} from "./types.js";

/**
 * Generate a unique subtask ID
 */
export const generateSubtaskId = (taskId: string, index: number): string =>
  `${taskId}-sub-${String(index + 1).padStart(3, "0")}`;

/**
 * Heuristics for detecting if a task needs decomposition
 */
export interface DecompositionHeuristics {
  /** Task mentions multiple files/components */
  hasMultipleTargets: boolean;
  /** Task has multiple distinct actions (add X, update Y, test Z) */
  hasMultipleActions: boolean;
  /** Task description is long (>500 chars) */
  isComplex: boolean;
  /** Task mentions testing explicitly */
  requiresTesting: boolean;
  /** Task mentions documentation */
  requiresDocs: boolean;
}

/**
 * Analyze a task to determine decomposition needs
 */
export const analyzeTask = (task: Task): DecompositionHeuristics => {
  const text = `${task.title} ${task.description || ""}`.toLowerCase();
  
  // Check for multiple file/component mentions
  const filePatterns = /\b(file|component|module|service|class|function|test|spec)\b/gi;
  const fileMatches = text.match(filePatterns) || [];
  const hasMultipleTargets = fileMatches.length > 2;
  
  // Check for multiple action words
  const actionPatterns = /\b(add|create|update|modify|fix|remove|delete|implement|refactor|test|document)\b/gi;
  const actionMatches = text.match(actionPatterns) || [];
  const uniqueActions = new Set(actionMatches.map(a => a.toLowerCase()));
  const hasMultipleActions = uniqueActions.size > 2;
  
  // Check complexity by length
  const isComplex = (task.description?.length || 0) > 500;
  
  // Check for explicit testing requirement (match plurals too)
  const requiresTesting = /\b(tests?|specs?|coverage|verify|validate|unit test|e2e)\b/i.test(text);
  
  // Check for documentation requirement (match plurals and variations)
  const requiresDocs = /\b(docs?|documentation|readme|comments?|jsdoc|tsdoc)\b/i.test(text);
  
  return {
    hasMultipleTargets,
    hasMultipleActions,
    isComplex,
    requiresTesting,
    requiresDocs,
  };
};

/**
 * Simple rule-based decomposition for common patterns
 */
export const decomposeByRules = (task: Task): Subtask[] => {
  const heuristics = analyzeTask(task);
  const subtasks: Subtask[] = [];
  let index = 0;
  
  // If task is simple, just create one subtask
  if (!heuristics.hasMultipleActions && !heuristics.isComplex && !heuristics.hasMultipleTargets) {
    return [
      {
        id: generateSubtaskId(task.id, 0),
        description: `${task.title}\n\n${task.description || ""}`.trim(),
        status: "pending",
      },
    ];
  }
  
  // For complex tasks, break into logical phases
  
  // Phase 1: Implementation
  subtasks.push({
    id: generateSubtaskId(task.id, index++),
    description: `Implement: ${task.title}\n\n${task.description || ""}`,
    status: "pending",
  });
  
  // Phase 2: Testing (if not explicitly part of implementation)
  if (heuristics.requiresTesting && !task.title.toLowerCase().includes("test")) {
    subtasks.push({
      id: generateSubtaskId(task.id, index++),
      description: `Add tests for: ${task.title}\n\nVerify the implementation works correctly with unit tests.`,
      status: "pending",
    });
  }
  
  // Phase 3: Documentation (if required)
  if (heuristics.requiresDocs) {
    subtasks.push({
      id: generateSubtaskId(task.id, index++),
      description: `Document: ${task.title}\n\nAdd appropriate documentation/comments.`,
      status: "pending",
    });
  }
  
  return subtasks;
};

/**
 * Decompose a task into subtasks.
 * Uses heuristics for simple decomposition.
 * 
 * @param task - The task to decompose
 * @param options - Decomposition options
 * @returns Array of subtasks
 */
export const decomposeTask = (
  task: Task,
  options?: {
    /** Maximum number of subtasks to create */
    maxSubtasks?: number;
    /** Force single subtask (no decomposition) */
    forceSingle?: boolean;
  }
): Subtask[] => {
  const { maxSubtasks = 5, forceSingle = false } = options || {};
  
  if (forceSingle) {
    return [
      {
        id: generateSubtaskId(task.id, 0),
        description: `${task.title}\n\n${task.description || ""}`.trim(),
        status: "pending",
      },
    ];
  }
  
  const subtasks = decomposeByRules(task);
  
  // Limit to maxSubtasks
  return subtasks.slice(0, maxSubtasks);
};

/**
 * Read subtasks file for a task
 */
export const readSubtasks = (openagentsDir: string, taskId: string): SubtaskList | null => {
  const subtasksPath = getSubtasksPath(openagentsDir, taskId);
  if (!fs.existsSync(subtasksPath)) return null;
  
  try {
    const content = fs.readFileSync(subtasksPath, "utf-8");
    return JSON.parse(content) as SubtaskList;
  } catch {
    return null;
  }
};

/**
 * Write subtasks file for a task
 */
export const writeSubtasks = (openagentsDir: string, subtaskList: SubtaskList): void => {
  const subtasksDir = path.join(openagentsDir, "subtasks");
  if (!fs.existsSync(subtasksDir)) {
    fs.mkdirSync(subtasksDir, { recursive: true });
  }
  
  const subtasksPath = getSubtasksPath(openagentsDir, subtaskList.taskId);
  fs.writeFileSync(subtasksPath, JSON.stringify(subtaskList, null, 2));
};

/**
 * Update a subtask's status
 */
export const updateSubtaskStatus = (
  openagentsDir: string,
  taskId: string,
  subtaskId: string,
  status: Subtask["status"],
  error?: string
): SubtaskList | null => {
  const subtaskList = readSubtasks(openagentsDir, taskId);
  if (!subtaskList) return null;
  
  const subtask = subtaskList.subtasks.find(s => s.id === subtaskId);
  if (!subtask) return null;
  
  subtask.status = status;
  subtask.error = error;
  
  if (status === "in_progress") {
    subtask.startedAt = new Date().toISOString();
  } else if (status === "done") {
    subtask.completedAt = new Date().toISOString();
  } else if (status === "verified") {
    subtask.verifiedAt = new Date().toISOString();
  }
  
  subtaskList.updatedAt = new Date().toISOString();
  writeSubtasks(openagentsDir, subtaskList);
  
  return subtaskList;
};

/**
 * Create a new subtask list for a task
 */
export const createSubtaskList = (
  task: Task,
  options?: {
    maxSubtasks?: number;
    forceSingle?: boolean;
  }
): SubtaskList => {
  const now = new Date().toISOString();
  const subtasks = decomposeTask(task, options);
  
  return {
    taskId: task.id,
    taskTitle: task.title,
    subtasks,
    createdAt: now,
    updatedAt: now,
  };
};

/**
 * Get pending subtasks from a list
 */
export const getPendingSubtasks = (subtaskList: SubtaskList): Subtask[] =>
  subtaskList.subtasks.filter(s => s.status === "pending");

/**
 * Get the next subtask to work on
 */
export const getNextSubtask = (subtaskList: SubtaskList): Subtask | null => {
  // First, check for in_progress subtasks (resume)
  const inProgress = subtaskList.subtasks.find(s => s.status === "in_progress");
  if (inProgress) return inProgress;
  
  // Then, get first pending subtask
  const pending = subtaskList.subtasks.find(s => s.status === "pending");
  return pending || null;
};

/**
 * Check if all subtasks are complete
 */
export const isAllSubtasksComplete = (subtaskList: SubtaskList): boolean =>
  subtaskList.subtasks.every(s => s.status === "done" || s.status === "verified");

/**
 * Check if any subtask failed
 */
export const hasFailedSubtasks = (subtaskList: SubtaskList): boolean =>
  subtaskList.subtasks.some(s => s.status === "failed");

File: crates/agent/src/decompose.rs
//! Subtask Decomposition
//!
//! Breaks tasks into implementable subtasks to prevent "one-shot" failures.
//! Uses heuristics for simple decomposition.

use crate::error::AgentResult;
use crate::types::{Subtask, SubtaskList, SubtaskStatus, Task, get_subtasks_path};
use chrono::Utc;
use regex::Regex;
use std::fs;
use std::path::Path;

/// Generate a unique subtask ID
pub fn generate_subtask_id(task_id: &str, index: usize) -> String {
    format!("{}-sub-{:03}", task_id, index + 1)
}

/// Heuristics for detecting if a task needs decomposition
#[derive(Debug, Clone, Default)]
pub struct DecompositionHeuristics {
    /// Task mentions multiple files/components
    pub has_multiple_targets: bool,
    /// Task has multiple distinct actions (add X, update Y, test Z)
    pub has_multiple_actions: bool,
    /// Task description is long (>500 chars)
    pub is_complex: bool,
    /// Task mentions testing explicitly
    pub requires_testing: bool,
    /// Task mentions documentation
    pub requires_docs: bool,
}

/// Analyze a task to determine decomposition needs
pub fn analyze_task(task: &Task) -> DecompositionHeuristics {
    let description = task.description.as_deref().unwrap_or("");
    let text = format!("{} {}", task.title, description).to_lowercase();

    // Check for multiple file/component mentions
    let file_pattern =
        Regex::new(r"\b(file|component|module|service|class|function|test|spec)\b").unwrap();
    let file_matches: Vec<_> = file_pattern.find_iter(&text).collect();
    let has_multiple_targets = file_matches.len() > 2;

    // Check for multiple action words
    let action_pattern = Regex::new(
        r"\b(add|create|update|modify|fix|remove|delete|implement|refactor|test|document)\b",
    )
    .unwrap();
    let action_matches: std::collections::HashSet<_> = action_pattern
        .find_iter(&text)
        .map(|m| m.as_str().to_lowercase())
        .collect();
    let has_multiple_actions = action_matches.len() > 2;

    // Check complexity by length
    let is_complex = description.len() > 500;

    // Check for explicit testing requirement
    let requires_testing =
        Regex::new(r"\b(tests?|specs?|coverage|verify|validate|unit test|e2e)\b")
            .unwrap()
            .is_match(&text);

    // Check for documentation requirement
    let requires_docs =
        Regex::new(r"\b(docs?|documentation|readme|comments?|jsdoc|tsdoc)\b")
            .unwrap()
            .is_match(&text);

    DecompositionHeuristics {
        has_multiple_targets,
        has_multiple_actions,
        is_complex,
        requires_testing,
        requires_docs,
    }
}

/// Simple rule-based decomposition for common patterns
pub fn decompose_by_rules(task: &Task) -> Vec<Subtask> {
    let heuristics = analyze_task(task);
    let mut subtasks = Vec::new();
    let mut index = 0;

    let description = task.description.as_deref().unwrap_or("");

    // If task is simple, just create one subtask
    if !heuristics.has_multiple_actions && !heuristics.is_complex && !heuristics.has_multiple_targets
    {
        return vec![Subtask::new(
            generate_subtask_id(&task.id, 0),
            format!("{}\n\n{}", task.title, description).trim().to_string(),
        )];
    }

    // For complex tasks, break into logical phases

    // Phase 1: Implementation
    subtasks.push(Subtask::new(
        generate_subtask_id(&task.id, index),
        format!("Implement: {}\n\n{}", task.title, description),
    ));
    index += 1;

    // Phase 2: Testing (if not explicitly part of implementation)
    if heuristics.requires_testing && !task.title.to_lowercase().contains("test") {
        subtasks.push(Subtask::new(
            generate_subtask_id(&task.id, index),
            format!(
                "Add tests for: {}\n\nVerify the implementation works correctly with unit tests.",
                task.title
            ),
        ));
        index += 1;
    }

    // Phase 3: Documentation (if required)
    if heuristics.requires_docs {
        subtasks.push(Subtask::new(
            generate_subtask_id(&task.id, index),
            format!(
                "Document: {}\n\nAdd appropriate documentation/comments.",
                task.title
            ),
        ));
    }

    subtasks
}

/// Options for task decomposition
#[derive(Debug, Clone, Default)]
pub struct DecomposeOptions {
    /// Maximum number of subtasks to create
    pub max_subtasks: Option<usize>,
    /// Force single subtask (no decomposition)
    pub force_single: bool,
}

/// Decompose a task into subtasks
pub fn decompose_task(task: &Task, options: Option<DecomposeOptions>) -> Vec<Subtask> {
    let opts = options.unwrap_or_default();
    let max_subtasks = opts.max_subtasks.unwrap_or(5);

    if opts.force_single {
        let description = task.description.as_deref().unwrap_or("");
        return vec![Subtask::new(
            generate_subtask_id(&task.id, 0),
            format!("{}\n\n{}", task.title, description).trim().to_string(),
        )];
    }

    let subtasks = decompose_by_rules(task);

    // Limit to max_subtasks
    subtasks.into_iter().take(max_subtasks).collect()
}

/// Read subtasks file for a task
pub fn read_subtasks(openagents_dir: &str, task_id: &str) -> Option<SubtaskList> {
    let subtasks_path = get_subtasks_path(openagents_dir, task_id);
    let path = Path::new(&subtasks_path);

    if !path.exists() {
        return None;
    }

    let content = fs::read_to_string(path).ok()?;
    serde_json::from_str(&content).ok()
}

/// Write subtasks file for a task
pub fn write_subtasks(openagents_dir: &str, subtask_list: &SubtaskList) -> AgentResult<()> {
    let subtasks_dir = Path::new(openagents_dir).join("subtasks");
    if !subtasks_dir.exists() {
        fs::create_dir_all(&subtasks_dir)?;
    }

    let subtasks_path = get_subtasks_path(openagents_dir, &subtask_list.task_id);
    let content = serde_json::to_string_pretty(subtask_list)?;
    fs::write(&subtasks_path, content)?;

    Ok(())
}

/// Update a subtask's status
pub fn update_subtask_status(
    openagents_dir: &str,
    task_id: &str,
    subtask_id: &str,
    status: SubtaskStatus,
    error: Option<String>,
) -> AgentResult<Option<SubtaskList>> {
    let mut subtask_list = match read_subtasks(openagents_dir, task_id) {
        Some(list) => list,
        None => return Ok(None),
    };

    let subtask = match subtask_list.subtasks.iter_mut().find(|s| s.id == subtask_id) {
        Some(s) => s,
        None => return Ok(None),
    };

    subtask.status = status;
    subtask.error = error;

    let now = Utc::now().to_rfc3339();
    match status {
        SubtaskStatus::InProgress => subtask.started_at = Some(now.clone()),
        SubtaskStatus::Done => subtask.completed_at = Some(now.clone()),
        SubtaskStatus::Verified => subtask.verified_at = Some(now.clone()),
        _ => {}
    }

    subtask_list.updated_at = now;
    write_subtasks(openagents_dir, &subtask_list)?;

    Ok(Some(subtask_list))
}

/// Create a new subtask list for a task
pub fn create_subtask_list(task: &Task, options: Option<DecomposeOptions>) -> SubtaskList {
    let now = Utc::now().to_rfc3339();
    let subtasks = decompose_task(task, options);

    SubtaskList {
        task_id: task.id.clone(),
        task_title: task.title.clone(),
        subtasks,
        created_at: now.clone(),
        updated_at: now,
    }
}

/// Get pending subtasks from a list
pub fn get_pending_subtasks(subtask_list: &SubtaskList) -> Vec<&Subtask> {
    subtask_list
        .subtasks
        .iter()
        .filter(|s| s.status == SubtaskStatus::Pending)
        .collect()
}

/// Get the next subtask to work on
pub fn get_next_subtask(subtask_list: &SubtaskList) -> Option<&Subtask> {
    // First, check for in_progress subtasks (resume)
    if let Some(in_progress) = subtask_list
        .subtasks
        .iter()
        .find(|s| s.status == SubtaskStatus::InProgress)
    {
        return Some(in_progress);
    }

    // Then, get first pending subtask
    subtask_list
        .subtasks
        .iter()
        .find(|s| s.status == SubtaskStatus::Pending)
}

/// Check if all subtasks are complete
pub fn is_all_subtasks_complete(subtask_list: &SubtaskList) -> bool {
    subtask_list.subtasks.iter().all(|s| {
        s.status == SubtaskStatus::Done || s.status == SubtaskStatus::Verified
    })
}

/// Check if any subtask failed
pub fn has_failed_subtasks(subtask_list: &SubtaskList) -> bool {
    subtask_list
        .subtasks
        .iter()
        .any(|s| s.status == SubtaskStatus::Failed)
}

#[cfg(test)]
mod tests {
    use super::*;

    fn make_task(id: &str, title: &str, description: Option<&str>) -> Task {
        Task {
            id: id.to_string(),
            title: title.to_string(),
            description: description.map(|s| s.to_string()),
            status: None,
        }
    }

    #[test]
    fn test_generate_subtask_id() {
        assert_eq!(generate_subtask_id("task-1", 0), "task-1-sub-001");
        assert_eq!(generate_subtask_id("task-1", 9), "task-1-sub-010");
        assert_eq!(generate_subtask_id("task-1", 99), "task-1-sub-100");
    }

    #[test]
    fn test_analyze_simple_task() {
        let task = make_task("t1", "Fix a bug", Some("Simple fix"));
        let heuristics = analyze_task(&task);

        assert!(!heuristics.has_multiple_targets);
        assert!(!heuristics.has_multiple_actions);
        assert!(!heuristics.is_complex);
    }

    #[test]
    fn test_analyze_complex_task() {
        let task = make_task(
            "t2",
            "Add tests and docs",
            Some("Create tests for the module. Add documentation for all functions."),
        );
        let heuristics = analyze_task(&task);

        assert!(heuristics.requires_testing);
        assert!(heuristics.requires_docs);
    }

    #[test]
    fn test_decompose_simple_task() {
        let task = make_task("t1", "Fix typo", None);
        let subtasks = decompose_task(&task, None);

        assert_eq!(subtasks.len(), 1);
        assert!(subtasks[0].description.contains("Fix typo"));
    }

    #[test]
    fn test_decompose_complex_task() {
        let task = make_task(
            "t2",
            "Add feature X",
            Some("Add feature with tests and documentation required. This is a longer description that requires multiple components and modules to be updated."),
        );
        let subtasks = decompose_task(&task, None);

        // Should have implementation + tests + docs subtasks (or just more than 1)
        // The exact count depends on heuristics matching
        assert!(!subtasks.is_empty());
        // At minimum we should get the tests subtask since "tests" is mentioned
        let has_tests_subtask = subtasks.iter().any(|s| s.description.contains("tests"));
        assert!(has_tests_subtask || subtasks.len() == 1);
    }

    #[test]
    fn test_force_single_subtask() {
        let task = make_task(
            "t3",
            "Complex task",
            Some("Add tests and documentation for the complex feature"),
        );
        let opts = DecomposeOptions {
            force_single: true,
            ..Default::default()
        };
        let subtasks = decompose_task(&task, Some(opts));

        assert_eq!(subtasks.len(), 1);
    }

    #[test]
    fn test_max_subtasks_limit() {
        let task = make_task(
            "t4",
            "Big task",
            Some("Add tests, documentation, refactor code, fix bugs, update configs"),
        );
        let opts = DecomposeOptions {
            max_subtasks: Some(2),
            ..Default::default()
        };
        let subtasks = decompose_task(&task, Some(opts));

        assert!(subtasks.len() <= 2);
    }

    #[test]
    fn test_create_subtask_list() {
        let task = make_task("t5", "Test task", None);
        let list = create_subtask_list(&task, None);

        assert_eq!(list.task_id, "t5");
        assert_eq!(list.task_title, "Test task");
        assert!(!list.subtasks.is_empty());
    }

    #[test]
    fn test_get_next_subtask() {
        let task = make_task("t6", "Test", None);
        let list = create_subtask_list(&task, None);

        let next = get_next_subtask(&list);
        assert!(next.is_some());
        assert_eq!(next.unwrap().status, SubtaskStatus::Pending);
    }

    #[test]
    fn test_is_all_complete() {
        let task = make_task("t7", "Test", None);
        let mut list = create_subtask_list(&task, None);

        assert!(!is_all_subtasks_complete(&list));

        // Mark all as done
        for subtask in &mut list.subtasks {
            subtask.status = SubtaskStatus::Done;
        }

        assert!(is_all_subtasks_complete(&list));
    }
}
